### PR TITLE
Add python nosetest integration to ctest

### PR DESCRIPTION
**Pull request by johnwparent**

### CODE REVIEW

File: CMake/kwiver-setup-python.cmake
Comment (as6520) [inline CMake/kwiver-setup-python.cmake:212]: We should probably create a virtual environment here so that the test occurs in a new environment and then remove it once the test is complete.Maybe FIXTURES_CLEANUP can be used to remove the virtualenv https://cmake.org/cmake/help/latest/prop_test/FIXTURES_CLEANUP.html

**Review by as6520 — CHANGES_REQUESTED** [review]
Looks good for most of the part. However we should use a virtual environment for testing rather than installing the dependencies in the user site packages.

File: CMake/kwiver-setup-python.cmake
Comment (johnwparent) [reply CMake/kwiver-setup-python.cmake:212]: Good point, will make the edit.

**Comment by tao558** [comment]
Jenkins test this please.

File: CMake/dashboard-scripts/KWIVER_common.cmake
Comment (dstoup) [inline CMake/dashboard-scripts/KWIVER_common.cmake:375]: Isn't this line already in the 'finish_bindings.." PR? Assuming the intent is for this to land then to rebase that one?

File: CMake/kwiver-setup-python.cmake
Comment (dstoup) [inline CMake/kwiver-setup-python.cmake:210]: Minor spelling error "defiened"

File: CMake/kwiver-setup-python.cmake
Comment (dstoup) [inline CMake/kwiver-setup-python.cmake:232]: More minor spelling "sucessfull"

File: CMake/kwiver-setup-tests-python.cmake
Comment (dstoup) [inline CMake/kwiver-setup-tests-python.cmake:14]: Since this is output path, it may not matter but usually on WIN32, the bin/ directory contains Debug/Release .. or whichever configuration was selected. Not sure if that matters to capture here or not.

File: CMake/utils/kwiver-utils-tests-python.cmake
Comment (dstoup) [inline CMake/utils/kwiver-utils-tests-python.cmake:143]: The indentation and spacing in this function is a bit off.

File: python/kwiver/vital/tests/CMakeLists.txt
Comment (dstoup) [inline python/kwiver/vital/tests/CMakeLists.txt:56]: I don't think you need to predefine things that are empty like NOSE_RUNNER, kwiver_test_working_path, kwiver_test_working_path. I could be wrong depending on context but generally it's not required.

*Edit ... in the event that you need to explicitly set these so you can set up correct behaviour in kwiver-setup-tests-python, then I would suggest that you should revisit the design here. Best to avoid required side-effects.

File: python/requirements.txt
Comment (dstoup) [inline python/requirements.txt:10]: Should we consider setting a minimum version for these packages at least? In many ways, I'm in favor of making the version or version ranges explicit. It adds a bit of a maintenance headache because one will need to update it periodically, but on the other hand, it will prevent a hard to find error when these packages change in a way that breaks what we're doing. It's especially problematic, in my experience, with Python packages that seem to have a really short deprecation window before they yank functionality ... looking at you Numpy!

**Review by dstoup — COMMENTED** [review]
Nothing too serious in my comments. Mostly formatting issues that I started making on the finish_bindings PR. A few of the comments require a little thought though but my just be me not understanding something.

File: CMake/dashboard-scripts/KWIVER_common.cmake
Comment (johnwparent) [reply CMake/dashboard-scripts/KWIVER_common.cmake:375]: Exactly right.

File: python/requirements.txt
Comment (johnwparent) [reply python/requirements.txt:10]: I think version ranges are probably the way to go.

File: python/kwiver/vital/tests/CMakeLists.txt
Comment (johnwparent) [reply python/kwiver/vital/tests/CMakeLists.txt:56]: Those definitions were/are unnecessary, they will be removed.

File: CMake/kwiver-setup-tests-python.cmake
Comment (johnwparent) [reply CMake/kwiver-setup-tests-python.cmake:14]: That's something that should be captured, and something I had overlooked, thanks!

Comment (dstoup) [reply CMake/kwiver-setup-tests-python.cmake]: Having said what I said .. the problem is that, you need to use the generator expression,  $<CONFIGURATION> for visual studio which is known only at compile time not CMake config/generate time. So, it requires something like a custom_command and can be a real pain. I'm not sure how (if at all) we've addressed that in kwiver so far.

File: python/requirements.txt
Comment (dstoup) [reply python/requirements.txt]: Agree. Much better than a single version if that's possible Gives a bit of flexibility.

**Comment by johnwparent** [comment]
Jenkins test this please

**Comment by johnwparent** [comment]
Jenkins test this please.

**Comment by tao558** [comment]
Jenkins test this please

**Comment by tao558** [comment]
Jenkins test this please

**Comment by dstoup** [comment]
The CentoOS error is a failing test...
https://open.cdash.org/test/249992392

The Mac error is interesting. It looks like some process is trying to write to /Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/cv2 or otherwise access it without permissions. But the failure line does suggest write access error

https://open.cdash.org/viewBuildError.php?buildid=6767849

" File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pip/req/req_install.py", line 1064, in move_wheel_files
    isolated=self.isolated,
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pip/wheel.py", line 345, in move_wheel_files
    clobber(source, lib_dir, True)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pip/wheel.py", line 316, in clobber
    ensure_dir(destdir)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pip/utils/__init__.py", line 83, in ensure_dir
    os.makedirs(path)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/os.py", line 220, in makedirs
    mkdir(name, mode):

**Comment by as6520** [comment]
Based on the package path it looks like opencv

**Comment by as6520** [comment]
I haven't looked at the virtual environment logic yet. However one suspect would be pip trying to install packages from [requirements.txt](https://github.com/Kitware/kwiver/blob/master/python/requirements.txt) to the root installation.

**Comment by johnwparent** [comment]
The virtual env creation in centOS is failing because centOS's cmake does not have the ability to FindPython_3, so centOS isn't running the pip install and the tests fail with unmet dependencies.  

MacOS is failing because rather than from within the virtualenv, python gets resolved to be the system install and then Ameya is exactly right. There's an option in cmake specifically to prevent MacOS from doing that.

**Comment by dstoup** [comment]
> The virtual env creation in centOS is failing because centOS's cmake does not have the ability to FindPython_3, so centOS isn't running the pip install and the tests fail with unmet dependencies.
> 

What is the minimum CMake version required for this change? We currently have `cmake version 3.11.0`. If that's not sufficient then we need to update the `cmake_minimum_required` to a version that is acceptable.

> MacOS is failing because rather than from within the virtualenv, python gets resolved to be the system install and then Ameya is exactly right. There's an option in cmake specifically to prevent MacOS from doing that.

Good to hear that there is a solution. Mac can be a pain in that regard.

**Comment by johnwparent** [comment]
> What is the minimum CMake version required for this change? We currently have `cmake version 3.11.0`. If that's not sufficient then we need to update the `cmake_minimum_required` to a version that is acceptable.

> Good to hear that there is a solution. Mac can be a pain in that regard.

find_package(python3) was introduced in CMake 3.12, but the Mac specific virtualenv settings for that command (referenced above) were not introduced until CMake 3.16.

**Comment by dstoup** [comment]
You might be able to do something like the following in .travis/install-deps.sh

`wget https://cmake.org/files/v3.15/cmake-3.15.7-Linux-x86_64.sh
./cmake-3.15.7-Linux-x86_64.sh`

Possibly adding a chmod +x cmake-3.15.7-Linux-x86_64.sh in there as well. I don't actually see how/where cmake is getting installed so it must come with the travis instance but I believe would be version 3.12

**Comment by tao558** [comment]
Jenkins test this please

**Comment by johnwparent** [comment]
Jenkins test this please

**Comment by tao558** [comment]
Jenkins test this please

File: .travis/update-cmake.sh
Comment (tao558) [inline .travis/update-cmake.sh]: Is there a difference between this and $HOME/cmake-install defined in travis/install-deps.sh?

File: CMake/kwiver-setup-python.cmake
Comment (tao558) [inline CMake/kwiver-setup-python.cmake:260]: Nitpick: extra space after (

File: CMake/kwiver-setup-python.cmake
Comment (tao558) [inline CMake/kwiver-setup-python.cmake:262]: Same as above

File: CMake/kwiver-setup-python.cmake
Comment (tao558) [inline CMake/kwiver-setup-python.cmake:296]: Is this still TODO?

File: python/CMakeLists.txt
Comment (tao558) [inline python/CMakeLists.txt:3]: Spacing

File: python/kwiver/vital/tests/CMakeLists.txt
Comment (tao558) [inline python/kwiver/vital/tests/CMakeLists.txt:8]: Are this and descriptors supposed to be commented out?

File: python/kwiver/vital/tests/CMakeLists.txt
Comment (tao558) [inline python/kwiver/vital/tests/CMakeLists.txt:47]: I'm sorry I don't see it right away, but what is this doing?

**Review by tao558 — CHANGES_REQUESTED** [review]

File: .travis/update-cmake.sh
Comment (johnwparent) [reply .travis/update-cmake.sh]: This file was an attempt at a workaround to travis handling the user input request poorly. It's no longer used, this file has been deleted.

File: python/kwiver/vital/tests/CMakeLists.txt
Comment (johnwparent) [reply python/kwiver/vital/tests/CMakeLists.txt:8]: I believe we had a conversation about this a week or so ago and decided that because these tests caused failures related to the bindings that are fixed in #1077 that we should remove them from the testing suite in this branch. The lines should probably just be removed.

File: CMake/kwiver-setup-tests-python.cmake
Comment (tao558) [inline CMake/kwiver-setup-tests-python.cmake:33]: I could be wrong about this, but this looks like the only difference between the `if(win32)` branch and this `else`?

File: CMake/utils/kwiver-config-venv.cmake
Comment (tao558) [inline CMake/utils/kwiver-config-venv.cmake:5]: Some spacing differences here and a few lines down

File: CMake/kwiver-setup-python.cmake
Comment (johnwparent) [reply CMake/kwiver-setup-python.cmake:296]: No, and I moved away from that implementation anyway, so it's both out of date and misleading, I'll remove it.

**Comment by johnwparent** [comment]
Jenkins test this please

File: python/kwiver/vital/tests/CMakeLists.txt
Comment (tao558) [reply python/kwiver/vital/tests/CMakeLists.txt:8]: Yes, apologies for not remembering. Removing them is a good idea.

**Comment by johnwparent** [comment]
Jenkins test this please

File: CMake/kwiver-setup-tests-python.cmake
Comment (johnwparent) [reply CMake/kwiver-setup-tests-python.cmake:33]: You are correct, and more than likely the `if(VENV_CREATED)` logic can be moved outside the `if(win32)` branch, I've left it there for now because a Windows build w/ an updated CMake version hasn't completed yet, and the logic inside the venv_created section may not actually being identical.

File: python/kwiver/vital/tests/CMakeLists.txt
Comment (johnwparent) [reply python/kwiver/vital/tests/CMakeLists.txt:47]: That is a variable I missed when I was doing cleanup. Removed.

**Comment by johnwparent** [comment]
Jenkins test this please

**Comment by Erotemic** [comment]
I'm a bit late on this, but nose is deprecated: https://nose.readthedocs.io/en/latest/index.html

We should invest in using pytest instead.

**Comment by johnwparent** [comment]
@Erotemic I agree, and switching over to pytest is absolutely the plan that has been discussed going forward. This branch was intended to quickly add the already written nose based python test suites to the larger ctest run so we can test the python bindings for types and algos with the CI while both sets of bindings are still WIP. The plan was to have this branch merged in to provide testing for the bindings while the switch to pytest is done both in the CMake and binding tests. Of course this can change if you feel this is not the way to do it.

**Comment by Erotemic** [comment]
@johnwparent I do remember some conversation about that. As long as we haven't forgotten that pytest is the goal.

**Comment by johnwparent** [comment]
@Erotemic Absolutely. This branch is intended to be a very temporary solution, with the focus on getting the switch to pytest done ASAP.

File: CMake/activate_venv.bat.in
Comment (Erotemic) [inline CMake/activate_venv.bat.in:8]: Might be worth looking for the `$CONDA_EXE` environment variable and falling back on hoping that conda is in the PATH.

Comment (johnwparent) [reply CMake/activate_venv.bat.in:8]: That sounds like a much better direction. Will do.

File: python/CMakeLists.txt
Comment (tao558) [inline python/CMakeLists.txt:7]: So `py_pip_command` is just `python3_EXECUTABLE`, and `pip_args` is the upgrading args? It might be worth restructuring the variables in `CMake/kwiver-setup-python.cmake` to something like:

`set(pip_command "${python3_EXECUTABLE}" "-m" "pip")`

`set(pip_upgrade_command "${pip_command}" "-q" "install" "--upgrade" "pip")`

`set(pip_dep_install "${pip_command}" "-q"
                        "install"
                        "-r"
                        "${KWIVER_SOURCE_DIR}/python/requirements.txt"
                        "||"
                        "${CMAKE_COMMAND}" "-E" "echo"
                        "Pip install failed, consult build output. Python dependencies may not be met"`


Then in `python/CMakeLists.txt`:

```
# Python cmake, if testing enabled set pip to install test dependencies in venv
if (KWIVER_ENABLE_TESTS)
    if (PIP_COMMAND)
        python_target_add_command( "pip-install-upgrade"
                                   ${pip_upgrade_command}
                                   "Upgrading pip...")
        python_target_add_command( "pip-install-deps"
                                   ${pip_dep_install}
                                   "${pip_args}"
                                   "Pip installing Python dependencies..."
                                   "pip-install-upgrade")
    endif()
endif()
```

**Review by tao558 — CHANGES_REQUESTED** [review]

File: python/CMakeLists.txt
Comment (tao558) [inline python/CMakeLists.txt:4]: Are these going to be removed later?

File: python/CMakeLists.txt
Comment (tao558) [inline python/CMakeLists.txt:9]: Worth changing the name to something like `pip-install-deps`? Just don't want this getting confused with the literal command `pip`

File: python/CMakeLists.txt
Comment (johnwparent) [reply python/CMakeLists.txt:4]: Yup, thought I had committed that before I pushed.

File: python/CMakeLists.txt
Comment (tao558) [inline python/CMakeLists.txt:10]: Likewise with this variable name. Maybe `PIP_INSTALL_DEPS_COMMAND`?

File: python/CMakeLists.txt
Comment (johnwparent) [reply python/CMakeLists.txt:9]: Can only possibly be helpful.

**Review by tao558 — APPROVED** [review]

**Comment by johnwparent** [comment]
Jenkins test this please

File: CMake/activate_venv.sh.in
Comment (dstoup) [inline CMake/activate_venv.sh.in:8]: I don't understand the syntax here:
echo ${PATH%"$OLDPATH"} > conda_venv_path.txt

What is the % doing?

Also, i don't see these scripts being generated anywhere. Are they supposed to turn into a .sh/.bat file somewhere?

File: CMake/kwiver-setup-tests-python.cmake
Comment (dstoup) [inline CMake/kwiver-setup-tests-python.cmake:19]: is NOSE_COM == NOSE_COMMAND? If so, and very minor, I kind of prefer to avoid the abbreviation for clarity.

File: python/kwiver/vital/tests/CMakeLists.txt
Comment (dstoup) [inline python/kwiver/vital/tests/CMakeLists.txt:46]: I kind of like having the comments. Not sure why it's removed.

**Review by dstoup — COMMENTED** [review]
Nothing major ... mostly questions and minor nits.

File: CMake/activate_venv.sh.in
Comment (johnwparent) [reply CMake/activate_venv.sh.in:8]: The % is the bash character for remove suffix, which here is removing all entries to the path prior to virtualenv activation from the current path, leaving the location of the python executable in the virtualenv. 

These scripts are generated here: [here](https://github.com/Kitware/kwiver/pull/1076/files#diff-cdbd9b26619cb98e22c4253e99db29ab0757b328f44b005543d8c01aea6700d7R5)

File: python/requirements.txt
Comment (as6520) [inline python/requirements.txt:4]: If I recall correctly, there were some security vulnerabilities with pillow library in 6.2. I would recommend using 7.* for minimum version too

File: python/requirements.txt
Comment (johnwparent) [reply python/requirements.txt:1]: [This](https://github.com/Kitware/kwiver/pull/1076#discussion_r488678970) discussion is my primary reason. Those that don't have upper/lower bounds will receive them when I can determine an acceptable versioning range.

File: python/requirements.txt
Comment (johnwparent) [reply python/requirements.txt:15]: I don't think it actually is a dependency, and it definitely comes prepackaged with python. I can't remember why it was ever added, but it certainly can be removed, nice catch.

File: python/requirements.txt
Comment (as6520) [reply python/requirements.txt:1]: Ah I see thanks for the clarification!

File: CMake/activate_venv.sh.in
Comment (dstoup) [reply CMake/activate_venv.sh.in:8]: Ah, very cool! I did not know that, thanks! Regarding the script generation, it looks like I need to satisfy `if (CONDA)` before they will generate, which requires `if (Python3_INTERPRETER_ID STREQUAL "Anaconda")`. If that's correct then I guess that would explain why I'm not seeing them.

File: CMake/kwiver-setup-python.cmake
Comment (dstoup) [inline CMake/kwiver-setup-python.cmake:264]: Pretty minor, but you could probably use ${KWIVER_CMAKE_DIR}/utils here. That variable is defined in the top level CMakeLists.txt so should be available here. Small gains for future-proofing that location.

File: CMake/kwiver-setup-tests-python.cmake
Comment (dstoup) [inline CMake/kwiver-setup-tests-python.cmake:13]: Does the inclusion of $<CONFIG> work here? It's a generator expression and not set at configure time. Usually it requires a custom target or command to get it right but I can't see clearly if that's how this stuff is getting called or not.

**Review by dstoup — COMMENTED** [review]
Just a few more comments that I missed last time.

File: CMake/activate_venv.sh.in
Comment (johnwparent) [reply CMake/activate_venv.sh.in:8]: Right, we only need to use the scripts when we encounter conda, as the directory in which the virtualenvs are created can differ per installation, as opposed to the python module virtualenv, where the path can be explicitly specified upon creation. We run the scipts to capture the changes `conda activate env` makes to the path (and other env variables) so we can propagate them to our current CMake process context.

File: CMake/kwiver-setup-tests-python.cmake
Comment (johnwparent) [reply CMake/kwiver-setup-tests-python.cmake:13]: I was basing this variable on the CMake [docs](https://cmake.org/cmake/help/v3.15/manual/cmake-generator-expressions.7.html#id2) and eventually it is passed to a call to either a custom command, or a call to add test/test_properties all of which I believe perform the expansion of the generator expressions. I may have, of course, misunderstood.

Comment (dstoup) [reply CMake/kwiver-setup-tests-python.cmake:13]: That's fine ... that documentation is correct and $<CONFIG> is the right way to do what you are trying to do. As long as it's being used within a structure that runs during generation and is resolving correctly, all is well. I just couldn't easily follow the trail to verify that. Thanks!

File: python/requirements.txt
Comment (dstoup) [reply python/requirements.txt]: Correct, I recently updated python/requirements.txt to set
`pillow==7.1.0`
I did that because github is good about alerting when we have a deprecated or dangerous package/version. Github won't appreciate anything less than 7.1.0

File: CMake/kwiver-setup-tests-python.cmake
Comment (johnwparent) [reply CMake/kwiver-setup-tests-python.cmake:13]: I thought about it a little more, and while, with the current usage of kwiver-setup-python-tests, that genex is fine, however as we transition to pytest, the usage of this setup file will change, and the genex will cause problems. I'll rework a more robust solution.

**Comment by dstoup** [comment]
I believe this is ready. I will defer to Ameya at this point.

**Review by as6520 — APPROVED** [review]

**Comment by as6520** [comment]
LGTM

**Pull request merged**
